nemo/collections/tts/modules/fastpitch.py

```diff
@@ -245,17 +313,27 @@ def forward(
 
         # Calculate speaker embedding
         if self.speaker_emb is None or speaker is None:
-            spk_emb = 0
+            spk_emb = None
         else:
             spk_emb = self.speaker_emb(speaker).unsqueeze(1)
 
+        if self.speaker_encoder is not None:
+            spk_emb = self.speaker_encoder(
```

Comment:
@rlangman @redoctopus  We need your opinion on this. Currently speakers are represented as lookup table embeddings (Line 318), In Adapters project we represent speakers as GST Embeddings + Titanet embeddings + lookup table embeddings. Line 321 `self.speaker_encoder` is the class to represent speakers, this has GST, Titanet representation. We were thinking if we should include lookup table also in this speaker_encoder class https://github.com/NVIDIA/NeMo/blob/385d46acc15c7a48c25db9e8e1f0ccafb17fb2de/nemo/collections/tts/modules/submodules.py#L613. This will lead to just one speaker representation class, which will be much neater than having separate classes/methods and then somehow adding them.

This has only one down side -> the use of older models which do not use the SpeakerEncoder class to represent speakers but the lookup table was a part of the FastPitch class. To solve this problem, we either upload new versions of the model with SpeakerEncoder class to NGC or write a wrapper that will put the lookup table from older model checkpoints into the SpeakerEncoder class.

What do you think?

Reply:
We can raise a separate PR for this, I would like to not make this PR more complicated.

Reply:
After thinking about it, I don't think we should merge the speaker embeddings with the speaker encoder. The speaker encoder is very specific to these experiments with custom voice, and unlikely to be used by most other TTS problems.

But now that we are supporting utterance level titanet embeddings, we should really support all 3 workflows:

1. Condition only on lookup table.
2. Condition only on utterance level embeddings
3. Condition on both together using speaker encoder.

All 3 could be represented as instances of SpeakerEncoder, but just having a branching if-statement will be clearer.

Reply:
So how about we add a configuration for whether to learn the embedding (defaulting to True for backwards compatibility):

        if self.learn_speaker_embedding and n_speakers > 1:
            self.speaker_emb = torch.nn.Embedding(n_speakers, symbols_embedding_dim)
        else:
            self.speaker_emb = None

Then you can manually set it to false when you provide the speaker encoder. Then we can cleanly support all 3 common training workflows.

```
def get_speaker_embedding(self, speaker, reference_speaker_embedding, reference_spec, reference_spec_lens):
  if self.speaker_encoder:
    spk_emb = self.speaker_encoder(speaker=speaker, ...)
  elif self.speaker_emb:
    if speaker is None:
      raise ValueError()
    spk_emb = self.speaker_emb(speaker)
  elif reference_speaker_embedding:
    spk_emb = reference_speaker_embedding
  else:
    spk_emb = None

  return spk_emb
```

Reply:
I'm in favor of including the lookup table in the speaker encoder class if it makes things simpler and improves the abstraction/organization. A little pain in re-training in the short term is okay!

Reply:
Either way technically works. It looks like we all agree the existing SpeakerEncoder should have an embedding array internally.

If we don't want to go with my above suggestion and do want to have a SpeakerEncoder as the only means later down, then we should create and finalize an interface in this PR to avoid breaking the adapter models later.

Current SpeakerEncoder in submodules.py is the "spaghetti code" approach with tons of if statements. The clean way to implement it would be like:

```
class SpeakerEncoder(ABC):

  @abstractmethod
  def get_speaker_encoder(self, speaker: Tensor, reference_speaker_embedding: Tensor, ...) -> Tensor:
    raise NotImplementedError
```
        
Then have one implementation which is the current GST logic with one logic flow and the if/else statements removed. And another simple one:

```
class LookupTableSpeakerEncoder(SpeakerEncoder):
    def __init__(self, num_speakers):
      self.speaker_emb = torch.nn.Embedding(n_speakers, symbols_embedding_dim)

    def get_speaker_encoder(self, speaker: Tensor, ...):
      if speaker is None:
        raise ValueError()
     return self.speaker_emb(speaker)
``'

Reply:
Then to deprecate the old system we can have a branch that uses the existing lookup table when speaker encoder is not provided, and does not create a lookup table in the FP model when it is.

Then we take our time to retrain all of the FP models on NGC with the new LookupTableSpeakerEncoder and delete the old branch when we are ready.